### ISSUE TITLE
Website Generation doesn't automatically reflect new changes to checks.xml and checks/<type>/index.xml

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#17691]
I have read check documentation: https://checkstyle.org/checks/xxxxxx/nameofaffectedcheck.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

After modifying a Check’s summary (for example, AnnotationLocation), I observed the following:

<img width="1466" height="638" alt="Image" src="https://github.com/user-attachments/assets/3201be69-425b-48cb-955e-7318e97721af" />

After running `mvn clean verify`, the above change is reflected in `annotationlocation.xml` and `AnnotationLocationCheck.xml`

But it is not reflected in `checks/annotation/index.xml`, though `XdocsCategoryIndexTest`'s tests fails and gives following errors:

```
[ERROR]   XdocsCategoryIndexTest.testAllChecksListedInCategoryIndexAndDescriptionMatches:132 Check 'AnnotationLocation' in index 'src\site\xdoc\checks\annotation\index.xml': index description is not a prefix of XDoc description.                            
expected to start with: Checks location of annotation on language elements.                                                     
but was               : Checks location of annotation on language elements testing. By default, Check enforce to locate annotati
ons before target element, annotation should be located on separate line from target element. This check also verifies that the annotations are on the same indenting level as the annotated element if they are not on the same line.                          
[ERROR]   XdocsPagesTest.testAllModulesPageInSyncWithModuleSummaries:335->validateModulesSyncWithTheirSummaries:391 The summary 
for module AnnotationLocation in the file "src\site\xdoc\checks.xml" should match the first sentence of the "Description" subsection for this module in the file "annotationlocation.xml"                                                                       
expected: Checks location of annotation on language elements testing.                                                           
but was : Checks location of annotation on language elements.                                                                   
[INFO]
[ERROR] Tests run: 5597, Failures: 2, Errors: 0, Skipped: 53
```

( screenshot showing that the summary is not updated )
<img width="1869" height="886" alt="Image" src="https://github.com/user-attachments/assets/f9731571-8f4f-427d-8051-a658fffa3fb8" />

We need to manually place the new summary. Instead of this, `checks/annotation/index.xml` should be able to automatically fetch the newly modified summary.

After manually placing the new summary into the `index.xml` file and running `mvn clean verify` again, the `XdocsPagesTest.java` fails and gives this error:

```
The summary for check AnnotationLocation in the file "src\site\xdoc\checks\annotation\index.xml" should match the summary for this check in the file "src\site\xdoc\checks.xml"
expected: Checks location of annotation on language elements.
but was : Checks location of annotation on language elements testing.
Expected :Checks location of annotation on language elements.
Actual   :Checks location of annotation on language elements testing.
```

( screenshot showing that the summary is not updated )
<img width="1872" height="905" alt="Image" src="https://github.com/user-attachments/assets/6664c694-23f1-4c02-a27d-4fb5550e92cf" />

it is telling that [checks/annotation/index.xml](https://checkstyle.org/checks/annotation/index.html) and [checks.xml](https://checkstyle.org/checks.html) have different content. So to pass the test, we again need to manually put the new summary into `checks.xml` as well.

### Problem

- `checks/<type>/index.xml` does not automatically pull the updated summary from the check’s Javadoc.

- `checks.xml` also requires a manual update to stay in sync with the modified Javadoc.

### Expected Behavior

Both `checks/<type>/index.xml` and `checks.xml` should automatically fetch the updated summary from the check’s Javadoc, similar to how `annotationlocation.xml` and `AnnotationLocationCheck.xml` are updated.

**Comment by nigar-08** [comment]
Hi ,
The issue is that checks/annotation/index.xml and checks.xml are not templates, so XdocGenerator doesn’t update them automatically when Javadoc changes in AnnotationLocationCheck.java. That’s why tests fail and manual edits are currently needed.
I can work on extending the generator so these files automatically fetch the updated summaries from the Javadocs.

**Comment by Zopsss** [comment]
> The issue is that checks/annotation/index.xml and checks.xml are not templates, so XdocGenerator doesn’t update them automatically when Javadoc changes in AnnotationLocationCheck.java.

Yes, that maybe the reason why we need to update docs manually.

Thank you for your interest in this issue. But if you're a beginner then we would recommend you to work on `good-*` issues first as this can be a bit tricky for you :) 

After gaining some experience first, you can work on this issue.

**Comment by nigar-08** [comment]
I completely understand—I'll start with some good-* issues to get more familiar first.
If there are any specific beginner-friendly issues you’d recommend for me, I’d love to take a look at those too!

**Comment by Zopsss** [comment]
Please start from good-first-issues

**Comment by romani** [comment]
@nigar-08 ,
Please [read](https://github.com/checkstyle/checkstyle/blob/master/.github/CONTRIBUTING.md#getting-started) and watch videos at [Starting_Development](https://checkstyle.org/beginning_development.html#Starting_Development).

**Comment by smita1078** [comment]
@Zopsss 

Either we can Update the existing generator logic to automatically sync the latest Javadoc summary into checks.xml and category index files during doc generation.
Or  we can Refactor to use a macro-based approach where summaries are fetched dynamically from the Javadoc at build time, the main idea would be to convert the existing static XML files into templates so they can dynamically pull summaries from the Check’s Javadoc.

which one should be prefer I can give it a try..

**Comment by Zopsss** [comment]
I think macro-based approach would be more cleaner. @romani wdyt?

**Comment by romani** [comment]
Macro based approach is better. Single special macro based that generate xml for section based on on package name.
https://github.com/checkstyle/checkstyle/blob/67b17c3ec85693705e0743a4d7112fa1a0017ee9/src/site/xdoc/checks/sizes/index.xml#L9

So test execution, will generate from template files the final xdoc xml with all content that is required.
Validation will not be required after implementation.

**Comment by romani** [comment]
Final fix is merged

**Comment by romani** [comment]
@smita1078 , thanks a lot for your hard work!!!

**Comment by smita1078** [comment]
> [@smita1078](https://github.com/smita1078) , thanks a lot for your hard work!!!

Thankyou!

**Comment by smita1078** [comment]
We need to Re-Open this, As currently fix is merged for generation of Checks.xml only..
I see there are two parts to this issue.
Website Generation doesn't automatically reflect new changes to checks.xml and checks/<type>/index.xml
1. checks.xml (merged)
2. checks/<type>/index.xml (yet to be done)

**Comment by smita1078** [comment]
### Context
I'm working on the next part where macro auto-generates `index.xml` files for each check category.

### Issue
Different categories use different title conventions in their existing `index.xml` files:

- **Simple plural**: `annotation` → "Annotations", `modifier` → "Modifiers"
- **Add "Checks"**: `blocks` → "Block Checks", `design` → "Design Checks"  
- **Keep singular**: `coding` → "Coding", `whitespace` → "Whitespace"
- **Expand/rename**: 
  - `misc` → "Miscellaneous"
  - `imports` → "Import Control"
  - `javadoc` → "Javadoc Comments"
  - `naming` → "Naming Conventions"
  - `regexp` → "Regular Expressions"
  - `sizes` → "Size Violations"

Which approach do you prefer for the automated index generation:

**Option 1: Single Template with Title Mapping**
- One template file: `checks-index.xml.template`
- Add a title mapping in the code:
```java
private static final Map<String, String> CATEGORY_TITLES = Map.ofEntries(
    Map.entry("annotation", "Annotations"),
    Map.entry("blocks", "Block Checks"),
    Map.entry("coding", "Coding"),
    // ... etc
);
```

**Option 2: Per-Category Templates**
- Multiple templates: `annotation-index.xml.template`, `blocks-index.xml.template`, etc.
- Each has its own title and structure

**Option 3: Standardize All Titles**
- Make all categories follow one pattern (e.g., "Category Name Checks")
- Update existing index.xml files to match

Which approach aligns better with the project's goals? @romani @Zopsss

**Comment by romani** [comment]
Option 1

**Comment by Zopsss** [comment]
@smita1078 other than the Page titles, does all the pages follow same structure? Or some have different structure? If only "title" is the issue then we can go with the first approach. But if there are some other cases as well then please let me know. We should discuss about them first.

**Comment by romani** [comment]
https://checkstyle.org/checks.html is done by macros
https://github.com/checkstyle/checkstyle/blob/6ea4dcbe97b9b3f31f675c70482e7dee412c120a/src/site/xdoc/checks.xml.template#L28-L33

Pages to be macrosed:
https://checkstyle.org/checks/annotation/index.html
https://checkstyle.org/checks/blocks/index.html
....
https://checkstyle.org/checks/javadoc/index.html
....

respective xml files that needs macros:
https://github.com/checkstyle/checkstyle/blob/master/src/site/xdoc/checks/annotation/index.xml
https://github.com/checkstyle/checkstyle/blob/master/src/site/xdoc/checks/blocks/index.xml
https://github.com/checkstyle/checkstyle/blob/master/src/site/xdoc/checks/javadoc/index.xml

I would simply make  macros for list of checks summaries, all other titles should stay in page template as hard coded.

Macros just generate list of summaries by group.
We have macros that does all Checks, now we need extension of it to print only by defined package name

**Comment by smita1078** [comment]
> [@smita1078](https://github.com/smita1078) other than the Page titles, does all the pages follow same structure? Or some have different structure? If only "title" is the issue then we can go with the first approach. But if there are some other cases as well then please let me know. We should discuss about them first.

Current Inconsistency:
annotation/index.xml (current):

Title: "Annotations"
Section name: "Annotation Checks"

All other categories follow a consistent pattern:

blocks: Title "Blocks" → Section "Blocks Checks"
coding: Title "Coding" → Section "Coding Checks"
design: Title "Design" → Section "Design Checks"
misc: Title "Miscellaneous" → Section "Miscellaneous Checks"

The Pattern:
Every category except annotation follows this rule:

Page title: [Category Name]
Section heading: [Same Category Name] + " Checks"

The annotation category is the only one that uses:

Page title: "Annotations" (plural)
Section heading: "Annotation Checks" (singular + "Checks")

**Comment by romani** [comment]
Please align naming

**Comment by Zopsss** [comment]
@smita1078 the annotations category should not be a problem right? Because we're maintaining hash map. 


Please proceed with the first approach.

**Comment by smita1078** [comment]
> [@smita1078](https://github.com/smita1078) the annotations category should not be a problem right? Because we're maintaining hash map.

Yes

I wanted clarification on this, there should be one template in the root folder or same template should be there in every check folder example one for annotation then one for blocks.. because keeping one in the root is generating extra unnecessary file copy of template file during mvn build

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/master/src/site/xdoc/checks/javadoc/index.xml

should be like:
```
<?xml version="1.0" encoding="UTF-8"?>
<document xmlns="http://maven.apache.org/XDOC/2.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/XDOC/2.0 https://maven.apache.org/xsd/xdoc-2.0.xsd">
  <head>
    <title>Javadoc Comments</title>
  </head>
  <body>
    <section name="Javadoc Comments Checks">
      <div class="wrapper">
          <macro name="allCheckSummaries" package="javadoc" />
      </div>
    </section>
  </body>
</document>
```

I think table tag should be part of macro, not a template. Macro should generate whole tag.

**Comment by romani** [comment]
Final fix is merged!
No more copy paste.
